backend/pkg/modules/internal/analytics_core/init.go

```diff
@@ -46,16 +55,42 @@ type AnalyticsCore struct {
 	info                *core.ModuleInfo
 	numWorkers          int
 	proccFuncRegistered map[TopicType][]AnalyticsModuleProccFunc
+	dataFramesRegistry  DataFramesRegistry
 	topics              []TopicType
 }
 
-type ProcFuncDataFrames struct {
-	dataFrames map[int]*interface{}
+type DataFramesRegistry struct {
+	dataFrames map[DataFrameID]DataFrame
+}
+
+func NewDataFramesRegistry(numWorkers int) DataFramesRegistry {
+	return DataFramesRegistry{
+		dataFrames: map[DataFrameID]DataFrame{},
+	}
+}
+
+func (dfr DataFramesRegistry) NewDataFrame(name DataFrameID, shards int) error {
+	dfr.dataFrames[name] = make(map[int]dataframe.DataFrame)
+	for i := 0; i < shards; i++ {
+		df := cache.DataFrame{}
+		err := df.Init(defaultDataFrameTTL)
+		if err != nil {
+			return fmt.Errorf("unable to create shard %d of dataframe %s", i, name)
+		}
+		dfr.dataFrames[name][i] = &df
+	}
+
+	return nil
+}
+
+func (dfr DataFramesRegistry) Get(name DataFrameID) (DataFrame, bool) {
+	df, found := dfr.dataFrames[name]
+	return df, found
 }
 
 type AnalyticsModuleProccFunc interface {
 	GetPriority() int
-	ProccFunc(topicName TopicType, dataFrames *ProcFuncDataFrames, partitionID int, message pubsub.MessageForBroker, annotations []interface{}, handler *AnalyticsCore) (newAnnotations []interface{})
+	ProccFunc(topicName TopicType, dataFramesRegistry DataFramesRegistry, partitionID int, message pubsub.MessageForBroker, annotations []interface{}, handler *AnalyticsCore) (newAnnotations []interface{})
```

Comment:
The only problem with such a design is that it is not safe, i.e it is possible to access data frames that belong to other workers and/or other shards... Meanwhile it is not critical and I prefer just to continue with that. 

By the way. Atleast the enforcement of the not sharing could easily be done. You can add to Get function the name of the processing function and also add it to the NewDataFrame...

It is not critical at this stage, I think
..